File: python/prism_sdk/domain_evidence_intake.py
"""Typed models for exact-digest raw domain evidence intake."""

from __future__ import annotations

from dataclasses import dataclass
import json
from typing import Any, Mapping, Sequence
from .artifacts import _digest, _mapping, _text
from .domain_reports import DOMAIN_REPORT_CLAIM_STATUSES, _bounded_text_list
from .errors import ArgumentError

DOMAIN_EVIDENCE_INTAKE_SCHEMA = "bioprism-devplat-domain-evidence-intake/0.1"
DOMAIN_EVIDENCE_INTAKE_WORKFLOW = "domain_evidence_intake"
DOMAIN_EVIDENCE_INTAKE_OUTCOMES = ("observed", "partial", "refused", "error", "unknown")
DOMAIN_EVIDENCE_INTAKE_COVERAGE_SCHEMA = "bioprism-devplat-domain-evidence-intake-coverage/0.1"
DOMAIN_EVIDENCE_INTAKE_COVERAGE_WORKFLOW = "domain_evidence_intake_coverage"

_MISSING = object()


def _json_value(name: str, value: Any) -> None:
    try:
        json.dumps(value, separators=(",", ":"), ensure_ascii=False)
    except (TypeError, ValueError) as error:
        raise ArgumentError(f"{name} must be JSON serializable") from error


@dataclass(frozen=True)
class DomainEvidenceIntakeRequest:
    """One caller-supplied raw request/response envelope."""

    group_id: str
    domains: tuple[str, ...]
    subject_id: str
    source_tool: str
    response: Any
    outcome: str
    claim_posture: Mapping[str, Any]
    request: Any = _MISSING
    parent_digests: tuple[str, ...] = ()
    source_plan_digest: str | None = None

    def __post_init__(self) -> None:
        _text("domain evidence intake group_id", self.group_id)
        _bounded_text_list("domain evidence intake domains", self.domains, required=True)
        _text("domain evidence intake subject_id", self.subject_id)
        _text("domain evidence intake source_tool", self.source_tool)
        if self.outcome not in DOMAIN_EVIDENCE_INTAKE_OUTCOMES:
            raise ArgumentError(
                "domain evidence intake outcome must be one of "
                + ", ".join(DOMAIN_EVIDENCE_INTAKE_OUTCOMES)
            )
        if not isinstance(self.claim_posture, Mapping):
            raise ArgumentError("domain evidence intake claim_posture must be an object")
        if self.claim_posture.get("status") not in DOMAIN_REPORT_CLAIM_STATUSES:
            raise ArgumentError("domain evidence intake claim_posture.status is invalid")
        _bounded_text_list(
            "domain evidence intake claim_posture.does_not_claim",
            self.claim_posture.get("does_not_claim"),
            required=True,
        )
        _bounded_text_list(
            "domain evidence intake claim_posture.limitations",
            self.claim_posture.get("limitations"),
        )
        _json_value("domain evidence intake response", self.response)
        if self.request is not _MISSING:
            _json_value("domain evidence intake request", self.request)
        if len(self.parent_digests) > 128:
            raise ArgumentError("domain evidence intake parent_digests must contain at most 128 values")
        for digest in self.parent_digests:
            _digest("domain evidence intake parent digest", digest)
        if self.source_plan_digest is not None:
            _digest("domain evidence intake source plan digest", self.source_plan_digest)

    @property
    def request_supplied(self) -> bool:
        return self.request is not _MISSING

    def to_arguments(self) -> dict[str, Any]:
        result: dict[str, Any] = {
            "group_id": self.group_id,
            "domains": list(self.domains),
            "subject_id": self.subject_id,
            "source_tool": self.source_tool,
            "response": self.response,
            "outcome": self.outcome,
            "claim_posture": dict(self.claim_posture),
            "parent_digests": list(self.parent_digests),
        }
        if self.source_plan_digest is not None:
            result["source_plan_digest"] = self.source_plan_digest
        if self.request is not _MISSING:
            result["request"] = self.request
        return result


@dataclass(frozen=True)
class DomainEvidenceIntakeReport:
    raw: dict[str, Any]
    group_id: str
    domains: tuple[str, ...]
    subject_id: str
    source_tool: str
    outcome: str
    source_plan_digest: str | None
    request_supplied: bool
    request_digest: str
    response_digest: str
    intake_digest: str
    report: Mapping[str, Any]
    intake: Mapping[str, Any]
    artifact_registry: Mapping[str, Any]
    catalogue_digest: str

    @classmethod
    def from_wire(cls, value: Mapping[str, Any]) -> "DomainEvidenceIntakeReport":
        raw = dict(value)
        if raw.get("workflow") != DOMAIN_EVIDENCE_INTAKE_WORKFLOW:
            raise ArgumentError("domain evidence intake workflow is invalid")
        if raw.get("schema") != DOMAIN_EVIDENCE_INTAKE_SCHEMA:
            raise ArgumentError("domain evidence intake schema is invalid")
        if raw.get("readiness_claimed") is not False:
            raise ArgumentError("domain evidence intake must not claim readiness")
        if raw.get("execution") != "not_started":
            raise ArgumentError("domain evidence intake execution must be not_started")
        request_supplied = raw.get("request_supplied")
        if not isinstance(request_supplied, bool):
            raise ArgumentError("domain evidence intake request_supplied must be a boolean")
        artifact_registry = _mapping("domain evidence intake artifact registry", raw.get("artifact_registry"))
        if artifact_registry.get("indexed") is not True:
            raise ArgumentError("domain evidence intake artifact registry projection is not indexed")
        domains = raw.get("domains")
        if not isinstance(domains, Sequence) or isinstance(domains, (str, bytes)):
            raise ArgumentError("domain evidence intake domains must be an array")
        intake = _mapping("domain evidence intake normalized envelope", raw.get("intake"))
        outcome = _text("domain evidence intake outcome", raw.get("outcome"))
        if outcome not in DOMAIN_EVIDENCE_INTAKE_OUTCOMES:
            raise ArgumentError("domain evidence intake outcome is invalid")
        return cls(
            raw=raw,
            group_id=_text("domain evidence intake group_id", raw.get("group_id")),
            domains=tuple(_text("domain evidence intake domain", domain) for domain in domains),
            subject_id=_text("domain evidence intake subject_id", raw.get("subject_id")),
            source_tool=_text("domain evidence intake source_tool", raw.get("source_tool")),
            outcome=outcome,
            source_plan_digest=(
                None
                if raw.get("source_plan_digest") is None
                else _digest("domain evidence intake source plan digest", raw.get("source_plan_digest"))
            ),
            request_supplied=request_supplied,
            request_digest=_digest("domain evidence intake request digest", raw.get("request_digest")),
            response_digest=_digest("domain evidence intake response digest", raw.get("response_digest")),
            intake_digest=_digest("domain evidence intake digest", raw.get("intake_digest")),
            report=_mapping("domain evidence intake report", raw.get("report")),
            intake=intake,
            artifact_registry=artifact_registry,
            catalogue_digest=_digest("domain evidence intake catalogue digest", raw.get("catalogue_digest")),
        )

    def to_dict(self) -> dict[str, Any]:
        return dict(self.raw)


@dataclass(frozen=True)
class DomainEvidenceIntakeCoverageRequest:
    group_id: str | None = None
    domain: str | None = None
    max_groups: int = 64
    include_intake_digests: bool = False

    def __post_init__(self) -> None:
        if self.group_id is not None:
            _text("domain evidence intake coverage group_id", self.group_id)
        if self.domain is not None:
            _text("domain evidence intake coverage domain", self.domain)
        if isinstance(self.max_groups, bool) or not isinstance(self.max_groups, int) or not 1 <= self.max_groups <= 128:
            raise ArgumentError("domain evidence intake coverage max_groups must be between 1 and 128")
        if not isinstance(self.include_intake_digests, bool):
            raise ArgumentError("domain evidence intake coverage include_intake_digests must be a boolean")

    def to_query_params(self) -> dict[str, str]:
        params = {
            "max_groups": str(self.max_groups),
            "include_intake_digests": str(self.include_intake_digests).lower(),
        }
        if self.group_id is not None:
            params["group_id"] = self.group_id
        if self.domain is not None:
            params["domain"] = self.domain
        return params

    def to_arguments(self) -> dict[str, Any]:
        return {
            "max_groups": self.max_groups,
            "include_intake_digests": self.include_intake_digests,
            **({"group_id": self.group_id} if self.group_id is not None else {}),
            **({"domain": self.domain} if self.domain is not None else {}),
        }


@dataclass(frozen=True)
class DomainEvidenceIntakeCoverageReport:
    raw: dict[str, Any]
    complete: bool
    group_count: int
    reported_group_count: int
    missing_group_count: int
    missing_group_ids: tuple[str, ...]
    tool_coverage_complete: bool
    missing_tool_group_ids: tuple[str, ...]
    domain_coverage_complete: bool
    missing_domain_group_ids: tuple[str, ...]
    groups: tuple[Mapping[str, Any], ...]
    catalogue_digest: str
    coverage_digest: str

    @classmethod
    def from_wire(cls, value: Mapping[str, Any]) -> "DomainEvidenceIntakeCoverageReport":
        raw = dict(value)
        if raw.get("workflow") != DOMAIN_EVIDENCE_INTAKE_COVERAGE_WORKFLOW:
            raise ArgumentError("domain evidence intake coverage workflow is invalid")
        if raw.get("schema") != DOMAIN_EVIDENCE_INTAKE_COVERAGE_SCHEMA:
            raise ArgumentError("domain evidence intake coverage schema is invalid")
        complete = raw.get("complete")
        if not isinstance(complete, bool):
            raise ArgumentError("domain evidence intake coverage complete must be a boolean")
        tool_coverage_complete = raw.get("tool_coverage_complete")
        if not isinstance(tool_coverage_complete, bool):
            raise ArgumentError("domain evidence intake tool_coverage_complete must be a boolean")
        domain_coverage_complete = raw.get("domain_coverage_complete")
        if not isinstance(domain_coverage_complete, bool):
            raise ArgumentError("domain evidence intake domain_coverage_complete must be a boolean")
        missing = _bounded_text_list(
            "domain evidence intake missing_group_ids", raw.get("missing_group_ids")
        )
        groups = raw.get("groups", [])
        if not isinstance(groups, Sequence) or isinstance(groups, (str, bytes)):
            raise ArgumentError("domain evidence intake coverage groups must be an array")
        return cls(
            raw=raw,
            complete=complete,
            group_count=_count("domain evidence intake group_count", raw.get("group_count")),
            reported_group_count=_count(
                "domain evidence intake reported_group_count", raw.get("reported_group_count")
            ),
            missing_group_count=_count(
                "domain evidence intake missing_group_count", raw.get("missing_group_count")
            ),
            missing_group_ids=missing,
            tool_coverage_complete=tool_coverage_complete,
            missing_tool_group_ids=_bounded_text_list(
                "domain evidence intake missing_tool_group_ids", raw.get("missing_tool_group_ids")
            ),
            domain_coverage_complete=domain_coverage_complete,
            missing_domain_group_ids=_bounded_text_list(
                "domain evidence intake missing_domain_group_ids", raw.get("missing_domain_group_ids")
            ),
            groups=tuple(_mapping("domain evidence intake coverage group", group) for group in groups),
            catalogue_digest=_digest(
                "domain evidence intake catalogue digest", raw.get("catalogue_digest")
            ),
            coverage_digest=_digest(
                "domain evidence intake coverage digest", raw.get("coverage_digest")
            ),
        )

    def to_dict(self) -> dict[str, Any]:
        return dict(self.raw)


def _count(name: str, value: Any) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or value < 0:
        raise ArgumentError(f"{name} must be a non-negative integer")
    return value

File: crates/devplat/src/domain_evidence_intake.rs
//! Bounded intake for raw tool outputs from every authoritative capability group.
//!
//! Intake is the bridge between an executed or externally supplied tool envelope and the
//! canonical domain-report/evidence layers. It records request and response JSON by exact digest,
//! retains the caller-declared outcome and claim posture, and refuses to infer meaning from a
//! response. Catalogue membership is enforced by the MCP boundary because the devplat crate does
//! not own the workspace's authoritative tool catalogue.

use crate::domain_report::{project_domain_report, validate_domain_report};
use bioprism_ids::ContentHash;
use serde_json::{json, Map, Value};
use std::collections::BTreeSet;
use thiserror::Error;

pub const DOMAIN_EVIDENCE_INTAKE_SCHEMA_VERSION: &str =
    "bioprism-devplat-domain-evidence-intake/0.1";
pub const DOMAIN_EVIDENCE_INTAKE_WORKFLOW: &str = "domain_evidence_intake";
pub const MAX_DOMAIN_EVIDENCE_INTAKE_BYTES: usize = 8 * 1024 * 1024;
pub const MAX_DOMAIN_EVIDENCE_INTAKE_DOMAINS: usize = 64;
pub const MAX_DOMAIN_EVIDENCE_INTAKE_PARENTS: usize = 128;
pub const MAX_DOMAIN_EVIDENCE_INTAKE_TEXT_BYTES: usize = 512;

const INTAKE_OUTCOMES: &[&str] = &["observed", "partial", "refused", "error", "unknown"];

#[derive(Debug, Clone, PartialEq, Eq, Error)]
pub enum DomainEvidenceIntakeError {
    #[error("domain evidence intake must be a JSON object")]
    NotObject,
    #[error("domain evidence intake field {0} is missing or invalid")]
    InvalidField(String),
    #[error("domain evidence intake field {field} exceeds the {maximum}-byte bound")]
    TextTooLarge { field: String, maximum: usize },
    #[error("domain evidence intake field {field} exceeds the {maximum}-item bound")]
    TooManyItems { field: String, maximum: usize },
    #[error("domain evidence intake is {actual} bytes, above the {maximum}-byte bound")]
    TooLarge { actual: usize, maximum: usize },
    #[error("domain evidence intake outcome must be one of: {0}")]
    InvalidOutcome(String),
    #[error("domain evidence intake report is invalid: {0}")]
    InvalidReport(String),
    #[error("domain evidence intake digest mismatch for {0}")]
    DigestMismatch(String),
    #[error("domain evidence intake could not be canonicalised: {0}")]
    Canonicalisation(String),
}

/// Normalize one raw request/response envelope into a canonical report-bearing intake record.
pub fn intake_domain_evidence(request: &Value) -> Result<Value, DomainEvidenceIntakeError> {
    let object = request
        .as_object()
        .ok_or(DomainEvidenceIntakeError::NotObject)?;
    ensure_size(request)?;

    let group_id = required_text(object, "group_id")?;
    let domains = text_array(object, "domains", MAX_DOMAIN_EVIDENCE_INTAKE_DOMAINS)?;
    if domains.is_empty() {
        return Err(DomainEvidenceIntakeError::InvalidField("domains".into()));
    }
    let subject_id = required_text(object, "subject_id")?;
    let source_tool = required_text(object, "source_tool")?;
    let response = object
        .get("response")
        .cloned()
        .ok_or_else(|| DomainEvidenceIntakeError::InvalidField("response".into()))?;
    let request_supplied = object.contains_key("request");
    let request_value = object.get("request").cloned().unwrap_or(Value::Null);
    let outcome = required_text(object, "outcome")?;
    if !INTAKE_OUTCOMES.contains(&outcome.as_str()) {
        return Err(DomainEvidenceIntakeError::InvalidOutcome(outcome));
    }
    let claim_posture = object
        .get("claim_posture")
        .cloned()
        .ok_or_else(|| DomainEvidenceIntakeError::InvalidField("claim_posture".into()))?;
    let parent_digests =
        digest_array(object, "parent_digests", MAX_DOMAIN_EVIDENCE_INTAKE_PARENTS)?;
    let request_digest = digest_value(&request_value)?;
    let response_digest = digest_value(&response)?;

    let observation = json!({
        "schema": DOMAIN_EVIDENCE_INTAKE_SCHEMA_VERSION,
        "workflow": DOMAIN_EVIDENCE_INTAKE_WORKFLOW,
        "group_id": group_id,
        "domains": domains,
        "subject_id": subject_id,
        "source_tool": source_tool,
        "request_supplied": request_supplied,
        "request": request_value,
        "response": response,
        "request_digest": request_digest,
        "response_digest": response_digest,
        "outcome": outcome
    });
    let intake_digest = digest_without_field(&observation, "intake_digest")?;
    let report = project_domain_report(&json!({
        "group_id": group_id,
        "domains": domains,
        "subject_id": subject_id,
        "source_tool": source_tool,
        "report": {"intake": observation},
        "claim_posture": claim_posture,
        "parent_digests": parent_digests
    }))
    .map_err(|error| DomainEvidenceIntakeError::InvalidReport(error.to_string()))?;

    let result = json!({
        "schema": DOMAIN_EVIDENCE_INTAKE_SCHEMA_VERSION,
        "workflow": DOMAIN_EVIDENCE_INTAKE_WORKFLOW,
        "group_id": group_id,
        "domains": domains,
        "subject_id": subject_id,
        "source_tool": source_tool,
        "request_supplied": request_supplied,
        "request_digest": request_digest,
        "response_digest": response_digest,
        "intake_digest": intake_digest,
        "outcome": outcome,
        "parent_digests": parent_digests,
        "report": report,
        "readiness_claimed": false,
        "execution": "not_started",
        "guarantees": [
            "request and response JSON are retained under exact canonical digests",
            "the canonical domain report preserves the caller-declared outcome and claim posture",
            "intake records a supplied envelope without executing or interpreting the source tool"
        ],
        "does_not_claim": [
            "a response or outcome label proves scientific, clinical, causal, or operational truth",
            "a source-tool membership check proves that the tool was executed or authorized",
            "intake proves provenance completeness, reproducibility, release readiness, or external effects"
        ]
    });
    ensure_size(&result)?;
    validate_domain_evidence_intake(&result)?;
    Ok(result)
}

/// Validate an intake artifact before registration, restore, or cross-process handoff.
pub fn validate_domain_evidence_intake(value: &Value) -> Result<(), DomainEvidenceIntakeError> {
    let object = value
        .as_object()
        .ok_or(DomainEvidenceIntakeError::NotObject)?;
    exact_text(object, "schema", DOMAIN_EVIDENCE_INTAKE_SCHEMA_VERSION)?;
    exact_text(object, "workflow", DOMAIN_EVIDENCE_INTAKE_WORKFLOW)?;
    let group_id = required_text(object, "group_id")?;
    let domains = text_array(object, "domains", MAX_DOMAIN_EVIDENCE_INTAKE_DOMAINS)?;
    if domains.is_empty() {
        return Err(DomainEvidenceIntakeError::InvalidField("domains".into()));
    }
    let subject_id = required_text(object, "subject_id")?;
    let source_tool = required_text(object, "source_tool")?;
    let request_supplied = object
        .get("request_supplied")
        .and_then(Value::as_bool)
        .ok_or_else(|| DomainEvidenceIntakeError::InvalidField("request_supplied".into()))?;
    let request_digest = required_digest(object, "request_digest")?;
    let response_digest = required_digest(object, "response_digest")?;
    let intake_digest = required_digest(object, "intake_digest")?;
    let outcome = required_text(object, "outcome")?;
    if !INTAKE_OUTCOMES.contains(&outcome.as_str()) {
        return Err(DomainEvidenceIntakeError::InvalidOutcome(outcome));
    }
    let parent_digests =
        digest_array(object, "parent_digests", MAX_DOMAIN_EVIDENCE_INTAKE_PARENTS)?;
    let report = object
        .get("report")
        .ok_or_else(|| DomainEvidenceIntakeError::InvalidField("report".into()))?;
    validate_domain_report(report)
        .map_err(|error| DomainEvidenceIntakeError::InvalidReport(error.to_string()))?;
    for (field, expected) in [
        ("group_id", group_id.as_str()),
        ("subject_id", subject_id.as_str()),
        ("source_tool", source_tool.as_str()),
    ] {
        if report.get(field).and_then(Value::as_str) != Some(expected) {
            return Err(DomainEvidenceIntakeError::InvalidField(format!(
                "report.{field}"
            )));
        }
    }
    if report.get("domains") != Some(&json!(domains)) {
        return Err(DomainEvidenceIntakeError::InvalidField(
            "report.domains".into(),
        ));
    }
    if report.get("parent_digests") != Some(&json!(parent_digests)) {
        return Err(DomainEvidenceIntakeError::InvalidField(
            "report.parent_digests".into(),
        ));
    }
    let observation = report
        .get("report")
        .and_then(Value::as_object)
        .and_then(|report| report.get("intake"))
        .and_then(Value::as_object)
        .ok_or_else(|| DomainEvidenceIntakeError::InvalidField("report.report.intake".into()))?;
    exact_text(observation, "schema", DOMAIN_EVIDENCE_INTAKE_SCHEMA_VERSION)?;
    exact_text(observation, "workflow", DOMAIN_EVIDENCE_INTAKE_WORKFLOW)?;
    if observation.get("group_id").and_then(Value::as_str) != Some(group_id.as_str())
        || observation.get("subject_id").and_then(Value::as_str) != Some(subject_id.as_str())
        || observation.get("source_tool").and_then(Value::as_str) != Some(source_tool.as_str())
        || observation.get("domains") != Some(&json!(domains))
    {
        return Err(DomainEvidenceIntakeError::InvalidField(
            "report.report.intake.identity".into(),
        ));
    }
    if observation.get("request_supplied") != Some(&Value::Bool(request_supplied))
        || observation.get("outcome").and_then(Value::as_str) != Some(outcome.as_str())
    {
        return Err(DomainEvidenceIntakeError::InvalidField(
            "report.report.intake.metadata".into(),
        ));
    }
    let request_value = observation.get("request").ok_or_else(|| {
        DomainEvidenceIntakeError::InvalidField("report.report.intake.request".into())
    })?;
    let response_value = observation.get("response").ok_or_else(|| {
        DomainEvidenceIntakeError::InvalidField("report.report.intake.response".into())
    })?;
    if digest_value(request_value)? != request_digest {
        return Err(DomainEvidenceIntakeError::DigestMismatch(
            "request_digest".into(),
        ));
    }
    if digest_value(response_value)? != response_digest {
        return Err(DomainEvidenceIntakeError::DigestMismatch(
            "response_digest".into(),
        ));
    }
    if observation.get("request_digest").and_then(Value::as_str) != Some(request_digest.as_str())
        || observation.get("response_digest").and_then(Value::as_str)
            != Some(response_digest.as_str())
    {
        return Err(DomainEvidenceIntakeError::DigestMismatch(
            "report.report.intake.digest".into(),
        ));
    }
    if digest_without_field(&observation.clone().into(), "intake_digest")? != intake_digest {
        return Err(DomainEvidenceIntakeError::DigestMismatch(
            "intake_digest".into(),
        ));
    }
    if object.get("readiness_claimed") != Some(&Value::Bool(false)) {
        return Err(DomainEvidenceIntakeError::InvalidField(
            "readiness_claimed".into(),
        ));
    }
    exact_text(object, "execution", "not_started")?;
    text_array(object, "guarantees", 16)?;
    text_array(object, "does_not_claim", 32)?;
    ensure_size(value)
}

fn digest_value(value: &Value) -> Result<String, DomainEvidenceIntakeError> {
    ContentHash::of_value(value)
        .map(|digest| digest.to_string())
        .map_err(|error| DomainEvidenceIntakeError::Canonicalisation(error.to_string()))
}

fn digest_without_field(value: &Value, field: &str) -> Result<String, DomainEvidenceIntakeError> {
    let mut object = value
        .as_object()
        .cloned()
        .ok_or(DomainEvidenceIntakeError::NotObject)?;
    object.remove(field);
    digest_value(&Value::Object(object))
}

fn required_digest(
    object: &Map<String, Value>,
    field: &str,
) -> Result<String, DomainEvidenceIntakeError> {
    let value = required_text(object, field)?;
    ContentHash::parse(value.clone())
        .map_err(|_| DomainEvidenceIntakeError::DigestMismatch(field.into()))?;
    Ok(value)
}

fn digest_array(
    object: &Map<String, Value>,
    field: &str,
    maximum: usize,
) -> Result<Vec<String>, DomainEvidenceIntakeError> {
    let Some(value) = object.get(field) else {
        return Ok(Vec::new());
    };
    let values = value
        .as_array()
        .ok_or_else(|| DomainEvidenceIntakeError::InvalidField(field.into()))?;
    let values = text_array_value(values, field, maximum)?;
    for value in &values {
        ContentHash::parse(value.clone())
            .map_err(|_| DomainEvidenceIntakeError::DigestMismatch(field.into()))?;
    }
    Ok(values)
}

fn text_array(
    object: &Map<String, Value>,
    field: &str,
    maximum: usize,
) -> Result<Vec<String>, DomainEvidenceIntakeError> {
    let values = object
        .get(field)
        .and_then(Value::as_array)
        .ok_or_else(|| DomainEvidenceIntakeError::InvalidField(field.into()))?;
    text_array_value(values, field, maximum)
}

fn text_array_value(
    values: &[Value],
    field: &str,
    maximum: usize,
) -> Result<Vec<String>, DomainEvidenceIntakeError> {
    if values.len() > maximum {
        return Err(DomainEvidenceIntakeError::TooManyItems {
            field: field.into(),
            maximum,
        });
    }
    let mut result = BTreeSet::new();
    for value in values {
        let text = value
            .as_str()
            .filter(|text| !text.trim().is_empty())
            .ok_or_else(|| DomainEvidenceIntakeError::InvalidField(field.into()))?;
        if text.len() > MAX_DOMAIN_EVIDENCE_INTAKE_TEXT_BYTES {
            return Err(DomainEvidenceIntakeError::TextTooLarge {
                field: field.into(),
                maximum: MAX_DOMAIN_EVIDENCE_INTAKE_TEXT_BYTES,
            });
        }
        result.insert(text.to_string());
    }
    Ok(result.into_iter().collect())
}

fn required_text(
    object: &Map<String, Value>,
    field: &str,
) -> Result<String, DomainEvidenceIntakeError> {
    let value = object
        .get(field)
        .and_then(Value::as_str)
        .filter(|value| !value.trim().is_empty())
        .ok_or_else(|| DomainEvidenceIntakeError::InvalidField(field.into()))?;
    if value.len() > MAX_DOMAIN_EVIDENCE_INTAKE_TEXT_BYTES {
        return Err(DomainEvidenceIntakeError::TextTooLarge {
            field: field.into(),
            maximum: MAX_DOMAIN_EVIDENCE_INTAKE_TEXT_BYTES,
        });
    }
    Ok(value.to_string())
}

fn exact_text(
    object: &Map<String, Value>,
    field: &str,
    expected: &str,
) -> Result<(), DomainEvidenceIntakeError> {
    if object.get(field).and_then(Value::as_str) != Some(expected) {
        return Err(DomainEvidenceIntakeError::InvalidField(field.into()));
    }
    Ok(())
}

fn ensure_size(value: &Value) -> Result<(), DomainEvidenceIntakeError> {
    let actual = serde_json::to_vec(value)
        .map_err(|error| DomainEvidenceIntakeError::Canonicalisation(error.to_string()))?
        .len();
    if actual > MAX_DOMAIN_EVIDENCE_INTAKE_BYTES {
        return Err(DomainEvidenceIntakeError::TooLarge {
            actual,
            maximum: MAX_DOMAIN_EVIDENCE_INTAKE_BYTES,
        });
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn request() -> Value {
        json!({
            "group_id": "biological_domains",
            "domains": ["modalities", "modalities"],
            "subject_id": "subject-1",
            "source_tool": "modality_catalog",
            "request": {"modality": "single_cell"},
            "response": {"modalities": ["single_cell"], "status": "bounded"},
            "outcome": "observed",
            "claim_posture": {
                "status": "observed",
                "does_not_claim": ["clinical validity"],
                "limitations": ["caller supplied envelope"]
            }
        })
    }

    #[test]
    fn intake_is_deterministic_and_preserves_exact_digests() {
        let first = intake_domain_evidence(&request()).unwrap();
        let second = intake_domain_evidence(&request()).unwrap();
        assert_eq!(first, second);
        assert_eq!(first["readiness_claimed"], false);
        assert_eq!(first["outcome"], "observed");
        assert_eq!(
            first["report"]["report"]["intake"]["request_supplied"],
            true
        );
        assert_eq!(first["request_digest"].as_str().unwrap().len(), 64);
        assert_eq!(first["response_digest"].as_str().unwrap().len(), 64);
        validate_domain_evidence_intake(&first).unwrap();
    }

    #[test]
    fn intake_supports_missing_request_and_explicit_refusal() {
        let mut value = request();
        value.as_object_mut().unwrap().remove("request");
        value["response"] = json!({"error": "caller refused execution"});
        value["outcome"] = json!("refused");
        let result = intake_domain_evidence(&value).unwrap();
        assert_eq!(result["request_supplied"], false);
        assert_eq!(result["outcome"], "refused");
        assert_eq!(result["report"]["report"]["intake"]["request"], Value::Null);
    }

    #[test]
    fn intake_refuses_invalid_outcomes_and_tampered_nested_digests() {
        let mut invalid = request();
        invalid["outcome"] = json!("success");
        assert!(matches!(
            intake_domain_evidence(&invalid),
            Err(DomainEvidenceIntakeError::InvalidOutcome(_))
        ));

        let mut tampered = intake_domain_evidence(&request()).unwrap();
        tampered["report"]["report"]["intake"]["response"] = json!({"tampered": true});
        assert!(matches!(
            validate_domain_evidence_intake(&tampered),
            Err(DomainEvidenceIntakeError::DigestMismatch(_))
        ));
    }
}
